tests/core/fsck-on-boot/task.yaml

```diff
@@ -0,0 +1,72 @@
+summary: the boot base provides essential fsck programs
+# TODO:UC20: this test really fails on core20, fsck is not invoked!
+systems: [-ubuntu-core-20-64]
+details: |
+  Snapd uses vfat on certain essential boot partitions, due to external
+  requirements imposed by the bootloader architecture. This test verifies that
+  the boot process is capable of detecting unclean fvat and fixing it before
+  such file system is mounted. This is an essential property to ensure
+  longevity of devices that rely on write to vfat to operate.
+prepare: |
+  tests.cleanup prepare
+execute: |
+  echo "We can corrupt the boot partition"
+  # FAT uses a specific byte to effectively indicate that the file system is
+  # dirty. The precise details as to how this byte is used by each system vary,
+  # but Linux sets it on a non-read-only mount, and clears it on unmount. We
+  # can set it manually, verify it when the image is mounted and observe fsck
+  # clearing it. Note that larger block devices use FAT32 and the offset
+  # differs. FAT12 and FAT16 uses 37 while FAT32 uses 65.
+  printf "\x01" > one
+  tests.cleanup defer rm -f one
+
+  cat /proc/self/mountinfo > mount.log
+
+  if [ "$SPREAD_REBOOT" -eq 0 ]; then
+    case "$SPREAD_SYSTEM" in
+      ubuntu-core-16-*)
+        # Refer to the core 16 PC gadget for details:
+        # https://github.com/snapcore/pc-amd64-gadget/blob/16/gadget.yaml
+        umount /boot/efi
+        # Use offset 64 as FAT32 kicks in for devices larger than 32MB
+        dd if=one of=/dev/sda2 seek=65 bs=1 count=1 conv=notrunc
+        ;;
+      ubuntu-core-18-*)
+        # Refer to the core 18 PC gadget for details:
+        # https://github.com/snapcore/pc-amd64-gadget/blob/18/gadget.yaml
+        umount /boot/efi
+        umount /boot/grub
+        # Use offset 64 as FAT32 kicks in for devices larger than 32MB
+        dd if=one of=/dev/sda2 seek=65 bs=1 count=1 conv=notrunc
```

Comment:
I've written a small tool back when we were investigating the corruption uboot.env when we got 2 files, one with uppercase and the other with lowercase names. The tool is here: https://gist.github.com/bboozzoo/e68254507eef4673dd8c6f9b82f65d92 would it make sense to use it as a starting point to try and corrupt the fs in a clever way, that would make it non-mountable?

Reply:
That's an interesting test but I wanted to prove something easier, that a dirty flag can be cleared because there is _some_ tool in the boot chain that handles that. This test is sufficient for that and doesn't require any more advanced corruption. I would love to add more comprehensive test but my point is this test is useful and (hopefully) precise in intent.